File: sol/solution/0001-0100/0017. Letter Combinations of a Phone Number/Solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        vector<string> mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> result;
        function<void(int, string)> backtrack = [&](int idx, string curr) {
            if (idx == digits.size()) {
                result.push_back(curr);
                return;
            }
            for (char c : mapping[digits[idx] - '0']) {
                backtrack(idx + 1, curr + c);
            }
        };
        backtrack(0, "");
        return result;
    }
};

File: sol/solution/0001-0100/0017. Letter Combinations of a Phone Number/Solution.ex
defmodule Solution do
  @spec letter_combinations(digits :: String.t) :: [String.t]
  def letter_combinations("") do
    []
  end
  def letter_combinations(digits) do
    mapping = %{
      "2" => ["a", "b", "c"],
      "3" => ["d", "e", "f"],
      "4" => ["g", "h", "i"],
      "5" => ["j", "k", "l"],
      "6" => ["m", "n", "o"],
      "7" => ["p", "q", "r", "s"],
      "8" => ["t", "u", "v"],
      "9" => ["w", "x", "y", "z"]
    }
    digits
    |> String.graphemes()
    |> Enum.map(&Map.get(mapping, &1, []))
    |> Enum.reduce([""], fn letters, acc ->
      for prefix <- acc, letter <- letters, do: prefix <> letter
    end)
  end
end